utils/buffer/src/lib.rs

```diff
@@ -1,3 +1,59 @@
+//! # `buffer_sv2`
+//!
+//! Handles memory management for Stratum v2 (Sv2) roles.
+//!
+//! Provides a memory-efficient buffer pool ([`BufferPool`]) that minimizes allocations and
+//! deallocations for high-throughput message frame processing in Sv2 roles. [`Slice`] helps
+//! minimize memory allocation overhead by reusing large buffers, improving performance and
+//! reducing latency. The [`BufferPool`] tracks the usage of memory slices, using atomic operations
+//! and shared state tracking to safely manage memory across multiple threads.
+//!
+//! ## Memory Structure
+//!
+//! The [`BufferPool`] manages a contiguous block of memory allocated on the heap, divided into
+//! fixed-size slots. Memory allocation within this pool operates in three distinct modes:
+//!
+//! 1. **Back Mode**: By default, memory is allocated sequentially from the back (end) of the
+//!    buffer pool. This mode continues until the back slots are fully occupied.
+//! 2. **Front Mode**: Once the back slots are exhausted, the [`BufferPool`] checks if any slots at
+//!    the front (beginning) have been freed. If available, it switches to front mode, allocating
+//!    memory from the front slots.
+//! 3. **Alloc Mode**: If both back and front slots are occupied, the [`BufferPool`] enters alloc
+//!    mode, where it allocates additional memory directly from the system heap. This mode may
+//!    introduce performance overhead due to dynamic memory allocation.
+//!
+//! [`BufferPool`] dynamically transitions between these modes based on slot availability,
+//! optimizing memory usage and performance.
+//!
+//! ## Usage
+//!
+//! When an incoming Sv2 message is received, it is buffered for processing. By default,
+//! memory is always first allocated from the back of the [`BufferPool`].
+//!
+//! If the all of the initially allocated buffer memory slots are completely filled and a new
+//! memory request comes in, [`BufferPool`] checks whether any memory has been freed at the back or
+//! the front using its shared state. If, for example, a slice has been freed that corresponds to
+//! the head of shared state, [`BufferPool`] will switch to front mode and start allocating
+//! incoming memory requests from there. However, if the entire shared state is full, it will
+//! switch to alloc mode and begin allocating system memory to fulfill incoming requests at a
+//! performance reduction.
+//!
+//! For subsequent memory requests, it will continue to check whether the prefix or suffix of the
+//! shared state has been freed. If it has, [`BufferPool`] will switch modes and start consuming
+//! pre-allocated memory; if not, it will remain in alloc mode.
+//!
+//! For isolated operations, [`Slice`] should manages its own memory ([`Vec<u8>`]). In
+//! high-performance applications, [`Slice`] should references externally managed memory from
+//! [`BufferPool`], reducing dynamic memory allocation.
```

Comment:
@Shourya742, with the new "Memory Structure" section added, what do you think of this for the Usage section:

```suggestion
//! ## Usage
//!
//! When an incoming Sv2 message is received, it is buffered for processing. The [`BufferPool`]
//! attempts to allocate memory from its internal slots, starting in back mode. If the back slots
//! are full, it checks for available front slots to switch to front mode. If no internal slots are
//! free, it resorts to alloc mode, allocating memory from the system heap.
//!
//! For operations requiring dedicated buffers, the [`Slice`] type manages its own memory using
//! [`Vec<u8>`]. In high-performance scenarios, [`Slice`] can reference externally managed memory
//! from the [`BufferPool`], reducing dynamic memory allocations and increasing performance.
```

Reply:
LGTM